api/src/main/java/io/druid/data/input/impl/prefetch/PrefetchableTextFilesFirehoseFactory.java

```diff
@@ -0,0 +1,248 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * PrefetchableTextFilesFirehoseFactory is an abstract firehose factory for reading text files.  The firehose returned
+ * by this class provides three key functionalities.
+ * <p/>
+ *
+ * - Caching: for the first call of {@link #connect(StringInputRowParser, File)}, it caches objects in a local disk
+ * up to maxCacheCapacityBytes.  These caches are NOT deleted until the process terminates, and thus can be used for
+ * future reads.
+ * <br/>
+ * - Fetching: when it reads all cached data, it fetches remaining objects into a local disk and reads data from
+ * them.  For the performance reason, prefetch technique is used, that is, when the size of remaining fetched data is
+ * smaller than {@link #prefetchTriggerBytes}, a background prefetch thread automatically starts to fetch remaining
+ * objects.
+ * <br/>
+ * - Retry: if an exception occurs while downloading an object, it retries again up to {@link #maxFetchRetry}.
+ * <p/>
+ *
+ * This implementation can be useful when the cost for reading input objects is large as reading from AWS S3 because
+ * batch tasks like IndexTask or HadoopIndexTask can read the whole data twice for determining partition specs and
+ * generating segments if the intervals of GranularitySpec is not specified.
+ * <br/>
+ * Prefetching can be turned on/off by setting maxFetchCapacityBytes.  Depending on prefetching is enabled or
+ * disabled, the behavior of the firehose is different like below.
+ * <p/>
+ *
+ * 1. If prefetch is enabled, this firehose can fetch input objects in background.
+ * <br/>
+ * 2. When next() is called, it first checks that there are already fetched files in local storage.
+ * <br/>
+ *   2.1 If exists, it simply chooses a fetched file and returns a {@link LineIterator} reading that file.
+ *   <br/>
+ *   2.2 If there is no fetched files in local storage but some objects are still remained to be read, the firehose
+ *   fetches one of input objects in background immediately. If an IOException occurs while downloading the object,
+ *   it retries up to the maximum retry count. Finally, the firehose returns a {@link LineIterator} only when the
+ *   download operation is successfully finished.
+ *   <br/>
+ * 3. If prefetch is disabled, the firehose returns a {@link LineIterator} which directly reads the stream opened by
+ * {@link #openObjectStream}. If there is an IOException, it will throw it and the read will fail.
+ */
+public abstract class PrefetchableTextFilesFirehoseFactory<ObjectType>
+    extends AbstractTextFilesFirehoseFactory<ObjectType>
+{
+  private static final Logger LOG = new Logger(PrefetchableTextFilesFirehoseFactory.class);
+  private static final long DEFAULT_MAX_CACHE_CAPACITY_BYTES = 1024 * 1024 * 1024; // 1GB
+  private static final long DEFAULT_MAX_FETCH_CAPACITY_BYTES = 1024 * 1024 * 1024; // 1GB
+  private static final long DEFAULT_FETCH_TIMEOUT = 60_000; // 60 secs
```

Comment:
`TimeUnits.SECONDS.toMillis(60)`. Also please add _MS suffix to the field name

Reply:
Fixed.